cardano-cli/src/Cardano/CLI/Common/Parsers.hs

```diff
@@ -6,22 +6,33 @@ module Cardano.CLI.Common.Parsers
 import           Cardano.Api (AnyConsensusModeParams (..), ConsensusModeParams (..),
                    EpochSlots (..), NetworkId (..), NetworkMagic (..), bounded)
 
+import           Cardano.CLI.Environment (maybeNetworkId)
 import           Data.Foldable
+import           Data.Maybe (maybeToList)
 import           Data.Word (Word64)
 import           Options.Applicative (Parser)
 import qualified Options.Applicative as Opt
 
 pNetworkId :: Parser NetworkId
-pNetworkId = asum
-  [ Opt.flag' Mainnet $ mconcat
-    [ Opt.long "mainnet"
-    , Opt.help "Use the mainnet magic id."
-    ]
-  , fmap (Testnet . NetworkMagic) $ Opt.option (bounded "TESTNET_MAGIC") $ mconcat
-    [ Opt.long "testnet-magic"
-    , Opt.metavar "NATURAL"
-    , Opt.help "Specify a testnet magic id."
+pNetworkId = asum $ mconcat
+  [ [ Opt.flag' Mainnet $ mconcat
+      [ Opt.long "mainnet"
+      , Opt.help $ mconcat
+        [ "Use the mainnet magic id. This overrides the CARDANO_NODE_NETWORK_ID "
+        , "environment variable"
+        ]
+      ]
+    , fmap (Testnet . NetworkMagic) $ Opt.option (bounded "TESTNET_MAGIC") $ mconcat
+      [ Opt.long "testnet-magic"
+      , Opt.metavar "NATURAL"
+      , Opt.help $ mconcat
+        [ "Specify a testnet magic id. This overrides the CARDANO_NODE_NETWORK_ID "
+        , "environment variable"
+        ]
+      ]
     ]
+  , -- Default to the network id specified by the environment variable if it is available.
+    pure <$> maybeToList maybeNetworkId
```

Comment:
`maybeNetworkid` which reads environment variables is very confusing here imo. Additionally, I don't think that any `Parser a` should do anything besides parsing CLI options. I think it would be better to move fallback to env variable somewhere else.

Reply:
The code that reads the environment variables is now in `main`.